File: src/server/bundleScore.js
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import path from 'node:path'

const BUNDLE_SCORE_VERSION = 1
const DEFAULTS = {
  extras_weight: 0.15,
  quality_gap_penalty: 0.05,
  fee_relief_scale: 6,
  extras_term_min: -8,
  extras_term_max: 12,
  score_version: BUNDLE_SCORE_VERSION,
}
const UNRANKED_KINDS = new Set(['near_haul', 'index_near_bundle'])

function clamp(value, low, high) {
  return Math.max(low, Math.min(high, Number(value)))
}

function scoreConfig(config) {
  const raw = (config && config.bundle_scoring) || {}
  const out = { ...DEFAULTS }
  for (const key of Object.keys(DEFAULTS)) {
    if (raw[key] != null) out[key] = raw[key]
  }
  const version = Number(raw.score_version)
  out.score_version = Number.isFinite(version) ? version : BUNDLE_SCORE_VERSION
  return out
}

function confidenceLabel(confidence) {
  if (confidence == null || !Number.isFinite(Number(confidence))) return null
  const value = Number(confidence)
  if (value < 0.6) return 'low'
  if (value < 0.8) return 'medium'
  return 'high'
}

function number(value) {
  if (typeof value === 'boolean') return null
  const n = Number(value)
  return Number.isFinite(n) ? n : null
}

function intScore(value) {
  const n = number(value)
  if (n == null) return null
  const asInt = Math.round(n)
  return asInt >= 0 && asInt <= 100 ? asInt : null
}

function isV2Member(member) {
  return Number(member?.score_version) === 2 && intScore(member?.buy_score) != null
}

function eligibleMembers(members) {
  const out = []
  for (const member of members || []) {
    if (!member || typeof member !== 'object') continue
    if (!isV2Member(member)) continue
    if (member.verification_concern === 'block') continue
    const buy = intScore(member.buy_score)
    const conf = number(member.score_confidence)
    if (buy == null || conf == null || conf < 0 || conf > 1) continue
    out.push({ ...member, buy_score: buy, score_confidence: conf })
  }
  return out
}

function isKeepMember(member, config) {
  if (String(member?.role || '').toLowerCase() === 'keep') return true
  const buy = intScore(member?.buy_score)
  const conf = number(member?.score_confidence)
  if (buy == null || conf == null) return false
  const keepMin = Number(config?.buy_scoring?.keep_min_score ?? 85)
  const minConf = Number(config?.buy_scoring?.min_keep_confidence ?? 0.6)
  return (
    buy >= keepMin &&
    conf >= minConf &&
    member.verification_concern !== 'block' &&
    member.hunt_fit !== false
  )
}

function nullResult(version) {
  return {
    bundle_score: null,
    bundle_score_version: version,
    bundle_confidence: null,
    bundle_anchor_item_id: null,
    bundle_rank_position: null,
  }
}

function splitAnchorExtras(eligible, anchor) {
  const extras = []
  let seenAnchor = false
  for (const member of eligible) {
    const same =
      member.buy_score === anchor.buy_score &&
      String(member.id) === String(anchor.id) &&
      member.score_confidence === anchor.score_confidence
    if (!seenAnchor && same) {
      seenAnchor = true
      continue
    }
    extras.push(member)
  }
  return extras
}

function calculateBundleScore(members, { kind = 'keep_bundle', config = null } = {}) {
  const cfg = scoreConfig(config)
  const version = Number(cfg.score_version)
  const kindKey = kind || 'keep_bundle'
  if (UNRANKED_KINDS.has(kindKey)) return nullResult(version)

  const eligible = eligibleMembers(members)
  if (!eligible.length) return nullResult(version)

  let anchorPool = eligible
  if (kindKey === 'keep_bundle' || kindKey === 'index_keep_bundle') {
    const keepPool = eligible.filter((m) => isKeepMember(m, config))
    anchorPool = keepPool.length ? keepPool : eligible
  }

  const anchor = [...anchorPool].sort((a, b) => {
    if (a.buy_score !== b.buy_score) return a.buy_score - b.buy_score
    if (a.score_confidence !== b.score_confidence) {
      return a.score_confidence - b.score_confidence
    }
    return String(a.id || '').localeCompare(String(b.id || ''))
  }).at(-1)

  const extras = splitAnchorExtras(eligible, anchor)
  const n = eligible.length
  const feeRelief = clamp(
    Number(cfg.fee_relief_scale) * ((n - 1) / n),
    0,
    Number(cfg.fee_relief_scale),
  )
  let extrasTerm = 0
  if (extras.length) {
    const extraMean =
      extras.reduce((sum, m) => sum + m.buy_score, 0) / extras.length
    const qualityGap = anchor.buy_score - extraMean
    extrasTerm = clamp(
      Number(cfg.extras_weight) * (extraMean - 50) -
        Number(cfg.quality_gap_penalty) * Math.max(qualityGap, 0) +
        feeRelief,
      Number(cfg.extras_term_min),
      Number(cfg.extras_term_max),
    )
  }

  const raw = clamp(anchor.buy_score + extrasTerm, 0, 100)
  const meanConf =
    eligible.reduce((sum, m) => sum + m.score_confidence, 0) / eligible.length
  const bundleConfidence = Math.min(anchor.score_confidence, meanConf)

  return {
    bundle_score: Math.round(raw),
    bundle_score_version: version,
    bundle_confidence: Number(bundleConfidence.toFixed(4)),
    bundle_anchor_item_id: anchor.id ?? null,
    bundle_rank_position: null,
  }
}

function applyToRow(row, config = null) {
  const out = { ...(row || {}) }
  const kind = out.kind || 'keep_bundle'
  Object.assign(out, calculateBundleScore(out.items || [], { kind, config }))
  return out
}

function fingerprint(row) {
  const sid = row?.seller_id
  const ids = (row?.items || [])
    .filter((it) => it && it.id != null)
    .map((it) => String(it.id))
    .sort()
  return `${sid}:${ids.join(',')}`
}

function assignBundleRanks(rows) {
  const decorated = (rows || []).map((row) => ({ ...(row || {}) }))
  const scored = []
  for (let index = 0; index < decorated.length; index += 1) {
    const score = intScore(decorated[index].bundle_score)
    if (score == null) {
      decorated[index].bundle_rank_position = null
      continue
    }
    scored.push({
      index,
      score,
      keptAt: String(decorated[index].kept_at || ''),
      fingerprint: fingerprint(decorated[index]),
    })
  }
  scored.sort((a, b) => a.fingerprint.localeCompare(b.fingerprint))
  scored.sort((a, b) => b.keptAt.localeCompare(a.keptAt))
  scored.sort((a, b) => b.score - a.score)
  scored.forEach((entry, position) => {
    decorated[entry.index].bundle_rank_position = position + 1
  })
  return decorated
}

function loadGoldenCases() {
  const here = path.dirname(fileURLToPath(import.meta.url))
  const goldenPath = path.join(
    here,
    '../../python/tests/fixtures/bundle_score_golden.json',
  )
  return JSON.parse(readFileSync(goldenPath, 'utf8'))
}

export {
  BUNDLE_SCORE_VERSION,
  UNRANKED_KINDS,
  scoreConfig,
  confidenceLabel,
  calculateBundleScore,
  applyToRow,
  assignBundleRanks,
  loadGoldenCases,
}

File: python/bundle_score.py
from __future__ import annotations

import math
from typing import Any

import buy_score

BUNDLE_SCORE_VERSION = 1
DEFAULTS = {
    "extras_weight": 0.15,
    "quality_gap_penalty": 0.05,
    "fee_relief_scale": 6.0,
    "extras_term_min": -8.0,
    "extras_term_max": 12.0,
}
UNRANKED_KINDS = frozenset({"near_haul"})


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, float(value)))


def score_config(config: dict | None = None) -> dict:
    raw = (config or {}).get("bundle_scoring") or {}
    out = dict(DEFAULTS)
    out.update({key: raw[key] for key in DEFAULTS if key in raw})
    try:
        version = int(raw.get("score_version", BUNDLE_SCORE_VERSION))
    except (TypeError, ValueError, OverflowError):
        version = BUNDLE_SCORE_VERSION
    out["score_version"] = version
    return out


def confidence_label(confidence: float | None) -> str | None:
    if confidence is None:
        return None
    value = float(confidence)
    if value < 0.60:
        return "low"
    if value < 0.80:
        return "medium"
    return "high"


def _number(value: Any) -> float | None:
    if isinstance(value, bool):
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if math.isfinite(number) else None


def _int_score(value: Any) -> int | None:
    number = _number(value)
    if number is None:
        return None
    as_int = int(round(number))
    return as_int if 0 <= as_int <= 100 else None


def _is_v2_member(member: dict) -> bool:
    return buy_score.is_v2_score(
        {
            "score_version": member.get("score_version"),
            "buy_score": member.get("buy_score"),
        }
    )


def _eligible_members(members: list[dict]) -> list[dict]:
    out = []
    for member in members:
        if not isinstance(member, dict):
            continue
        if not _is_v2_member(member):
            continue
        if member.get("verification_concern") == "block":
            continue
        buy = _int_score(member.get("buy_score"))
        if buy is None:
            continue
        conf = _number(member.get("score_confidence"))
        if conf is None or not 0 <= conf <= 1:
            continue
        row = dict(member)
        row["buy_score"] = buy
        row["score_confidence"] = conf
        out.append(row)
    return out


def _is_keep_member(member: dict, config: dict | None = None) -> bool:
    if str(member.get("role") or "").lower() == "keep":
        return True
    cfg = buy_score.score_config(config)
    buy = _int_score(member.get("buy_score"))
    conf = _number(member.get("score_confidence"))
    if buy is None or conf is None:
        return False
    return (
        buy >= int(cfg["keep_min_score"])
        and conf >= float(cfg["min_keep_confidence"])
        and member.get("verification_concern") != "block"
        and member.get("hunt_fit") is not False
    )


def _null_result(version: int) -> dict:
    return {
        "bundle_score": None,
        "bundle_score_version": version,
        "bundle_confidence": None,
        "bundle_anchor_item_id": None,
        "bundle_rank_position": None,
    }


def _split_anchor_extras(eligible: list[dict], anchor: dict) -> list[dict]:
    extras = []
    seen_anchor = False
    anchor_id = anchor.get("id")
    for member in eligible:
        same = (
            member["buy_score"] == anchor["buy_score"]
            and str(member.get("id")) == str(anchor_id)
            and member["score_confidence"] == anchor["score_confidence"]
        )
        if not seen_anchor and same:
            seen_anchor = True
            continue
        extras.append(member)
    return extras


def calculate_bundle_score(
    members: list[dict],
    *,
    kind: str | None = None,
    config: dict | None = None,
) -> dict:
    cfg = score_config(config)
    version = int(cfg["score_version"])
    kind_key = kind or "keep_bundle"
    if kind_key in UNRANKED_KINDS:
        return _null_result(version)

    eligible = _eligible_members(members or [])
    if not eligible:
        return _null_result(version)

    if kind_key in {"keep_bundle", "index_keep_bundle"}:
        keep_pool = [m for m in eligible if _is_keep_member(m, config)]
        anchor_pool = keep_pool or eligible
    else:
        anchor_pool = eligible

    anchor = max(
        anchor_pool,
        key=lambda m: (m["buy_score"], m["score_confidence"], str(m.get("id") or "")),
    )
    extras = _split_anchor_extras(eligible, anchor)
    n = len(eligible)
    fee_relief = clamp(
        cfg["fee_relief_scale"] * (n - 1) / n,
        0.0,
        float(cfg["fee_relief_scale"]),
    )
    if not extras:
        extras_term = 0.0
    else:
        extra_mean = sum(m["buy_score"] for m in extras) / len(extras)
        quality_gap = anchor["buy_score"] - extra_mean
        extras_term = clamp(
            float(cfg["extras_weight"]) * (extra_mean - 50.0)
            - float(cfg["quality_gap_penalty"]) * max(quality_gap, 0.0)
            + fee_relief,
            float(cfg["extras_term_min"]),
            float(cfg["extras_term_max"]),
        )

    raw = clamp(anchor["buy_score"] + extras_term, 0.0, 100.0)
    confidences = [m["score_confidence"] for m in eligible]
    bundle_confidence = min(
        anchor["score_confidence"],
        sum(confidences) / len(confidences),
    )

    return {
        "bundle_score": int(round(raw)),
        "bundle_score_version": version,
        "bundle_confidence": round(bundle_confidence, 4),
        "bundle_anchor_item_id": anchor.get("id"),
        "bundle_rank_position": None,
    }


def apply_to_row(row: dict, config: dict | None = None) -> dict:
    """Return a shallow copy of row with bundle score fields set."""
    out = dict(row or {})
    kind = out.get("kind") or "keep_bundle"
    out.update(
        calculate_bundle_score(out.get("items") or [], kind=kind, config=config)
    )
    return out


def _fingerprint(row: dict) -> str:
    sid = row.get("seller_id")
    ids = sorted(
        str(it.get("id"))
        for it in (row.get("items") or [])
        if isinstance(it, dict) and it.get("id") is not None
    )
    return f"{sid}:" + ",".join(ids)


def assign_bundle_ranks(rows: list[dict]) -> list[dict]:
    """Assign bundle_rank_position among scored rows; unscored stay null."""
    decorated = [dict(row or {}) for row in rows]
    scored: list[tuple[int, int, str, str]] = []
    for index, row in enumerate(decorated):
        score = _int_score(row.get("bundle_score"))
        if score is None:
            row["bundle_rank_position"] = None
            continue
        scored.append(
            (index, score, str(row.get("kept_at") or ""), _fingerprint(row))
        )
    scored.sort(key=lambda t: t[3])
    scored.sort(key=lambda t: t[2], reverse=True)
    scored.sort(key=lambda t: t[1], reverse=True)
    for position, (index, *_rest) in enumerate(scored, start=1):
        decorated[index]["bundle_rank_position"] = position
    return decorated
